### DISCUSSION TITLE
pg-boss getting started guide / examples

### DISCUSSION THREAD

**Discussion by brannan** [timgit/pg-boss#416] [General]
I've been programming on and off for a long time but somehow have never taken a deep dive into task queues. I did have one project that made extensive use of RabbitMQ, but I didn't do much of the work on that. So when it comes to pg-boss and task queues, I'm an idiot. I get the idea, certainly, but when to use send, publish, schedule took a little thought and digging around. Maybe this is the best way.

Your API docs are good, but but they miss the bigger picture for somebody who knows they need a task-queue but don't know the background theory and queuing terminology. For instance, "I want to offload the sending of messages to a queue. The messages will need to be sent at one specific time and date. Considerations and how to proceed?"

Anyway, I've dug through the docs, read a number of the discussions, and implemented the solution that I think I need. Well, it may not be best or even right, but it does work.

So I wrote a little "lay programmers guide to pg-boss" for my current project to document my understand of pg-boss. It's all of 57 lines long. 

Would this be of any benefit to the project? Somebody who knows more than I do will have to review it. And clearly this is not a barn burner of an issue.

Perhaps just pulling a few more examples into the first readme would be a good idea. I'm happy to take a shot at this if anybody thinks it would be useful.

**Comment by timgit** [comment]
I agree with you that the Intro section of the docs is not a holistic intro to job queues, but rather how to use the npm package.  This could definitely be improved. 

I'd prefer to keep this very small, however, and maybe link to someone else's intro, perhaps something in the AWS SQS docs since I reference that on other places. Something that mentions asynchronicity, persistence, retries, and idempotency would be a good start.

**Comment by 1mike12** [comment]
I cannot agree more with @brannan. I think this library may take 1st place for usefullness/impenetrable docs ratio for anything I've seen in the last few years. I know the library must be good with all the stars and great design, but without claude/gpt, I would have no idea how to interact with this without spending hours analyzing the source. Since so little is written about it, even the LLMs haven't had enough corpus to train on. 

There _is_ no other intro to this, except for an outdated article https://logsnag.com/blog/deep-dive-into-background-jobs-with-pg-boss-and-typescript

The responsibility of explaining how a library works to users lies with the library imo, not 3rd parties. 

### just a random list of questions that popped into my head within the last hour of getting started 

Does this library take advantage of postgres' notify feature? I mean it's built specifically for pg v13+, there would be no way this is skipped for polling only right?

what does Raw SQL support for non-Node.js runtimes via INSERT or COPY mean? 

it doesn't say how I have to setup my table for the library? Does it handle its own migration? what if I want to specify the table names?

what is the pubsub for and why/when would I use it vs the implicit calls?

how do I type the jobs? do I have to install a @typing ? does this have one?

What does the job object look like?

why is the job to be processed an array?

**Comment by infovore** [comment]
+1. @brannan , did you ever publish those 57 lines _anywhere_? I'd be grateful to see even somebody else's approach.

I both understand the sample code and cannot see, without having to do a lot of hammering myself, how to apply it in anything larger than a toy scenario. Should I, for instance, be always hitting the 'create queue' function in my worker script, and hope it just silently does nothing if the queue/tables already exist?

The sample code proves how things work, and doesn't really indicate anything about how to use it; at the very least, a sample of "separating out worker processing from pushing onto the queue" would be useful, given... that's mainly how you use a queue. 

I don't need a holistic introduction to queues. I do, however, need more than API docs - demonstrating _this_ library's preferred approach to common problems, its own philosophy explained. (The _last_ thing I want, ever, is a link to AWS documentation, which I almost universally loathe).

**Reply by olepbr** [reply]
> I do, however, need more than API docs - demonstrating this library's preferred approach to common problems, its own philosophy explained.

I would also very much appreciate some insight here. For instance: is the recommended way to use pg-boss to create one queue for each job type? This seems unfortunate to me, as (given that `boss.work()` operates on a single queue) the number of polling workers increases for every job type added. I guess the alternative is to encode the job type in the job payload and then letting the job handler/worker switch on that?

**Reply by timgit** [reply]
There's a request to bring back wildcard fetching/polling in #505.  It sounds like you'd be interested in that approach.

**Comment by hitsthings** [comment]
> Does this library take advantage of postgres' notify feature? I mean it's built specifically for pg v13+, there would be no way this is skipped for polling only right?

Anybody know the answer to this one? I see a`notifyWorker` method where you can notify a worker if you have their ID. Do I have to call that or is it automatic? If it's automatic, I could probably drastically lower my polling frequency and still handle jobs faster. That would be great.

Searching the source I don't see usages of worker.notify() or notifyWorker(id) so I am guessing it is not automatic. Also it looks like it would only work within the same process iiuc since it just looks up the worker in a local array. A coworker uses river-queue in Go which uses postgres NOTIFY to bridge that gap and immediate start a worker in a separate process/service when a job is available. That would be a killer addition to pg-boss.

**Comment by timgit** [comment]
There's a few historical threads on listen/notify here.  One is #93

**Comment by hitsthings** [comment]
Thanks @timgit !

**Comment by Abhishek21k** [comment]
try this also [sidequest](https://github.com/sidequestjs/sidequest)

**Comment by arikfr** [comment]
I think it will be very beneficial to provide a real world integration examples, where there are separate processes that produce and consume the tasks. This will help answers questions like: do we need to call `.start()` in every such process? Do we need to construct `PgBoss` with `supervise` false in non workers?

If time allows, I will try to deduce this from the source code and share back here.

**Reply by timgit** [reply]
Yes, you always need to run start().  The happy path is that you don't need to worry about supervise, schedule, or other options.  Even if you run 10 processes, all with supervise: true, only 1 of them will run supervision.
